packages/contracts/contracts/maci/Tally.sol

```diff
@@ -0,0 +1,321 @@
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed <= duration + cooldown) {
+      revert CooldownPeriodNotOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the voting period is over
+  modifier beforeVotingDeadline() {
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed > duration) {
+      revert VotingPeriodOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the voting period is not over
+  modifier afterVotingDeadline() {
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed <= duration) {
+      revert VotingPeriodNotOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the strategy is not initialized
+  modifier isInitialized() {
+    if (!initialized) {
+      revert NotInitialized();
+    }
+
+    _;
+  }
+
+  /// @notice Initialize tally with strategy params
+  /// @param params The strategy initialization params
+  function init(IPayoutStrategy.StrategyInit calldata params) public onlyOwner {
+    if (initialized) {
+      revert AlreadyInitialized();
+    }
+
+    if (params.poll != address(poll)) {
+      revert InvalidPoll();
+    }
+
+    initialized = true;
+    cooldown = params.cooldownTime;
+    registry = IPoll(params.poll).getRegistry();
+    token = IERC20(params.payoutToken);
+    maxContributionAmount = params.maxContribution;
+    voiceCreditFactor = params.maxContribution / MAX_VOICE_CREDITS;
+  }
+
+  /// @notice Pause contract calls (deposit, claim, withdraw)
+  function pause() public onlyOwner {
+    _pause();
+  }
+
+  /// @notice Unpause contract calls (deposit, claim, withdraw)
+  function unpause() public onlyOwner {
+    _unpause();
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function deposit(uint256 amount) public isInitialized whenNotPaused beforeVotingDeadline {
+    totalAmount += amount;
+
+    token.safeTransferFrom(msg.sender, address(this), amount);
+  }
+
+  /// @inheritdoc IPayoutStrategy
+  function withdrawExtra(
+    address[] calldata receivers,
+    uint256[] calldata amounts
+  ) public override isInitialized onlyOwner whenNotPaused afterCooldown {
+    uint256 amountLength = amounts.length;
+    uint256 totalFunds = totalAmount;
+    uint256 sum = 0;
+
+    for (uint256 index = 0; index < amountLength; ) {
+      uint256 amount = amounts[index];
+      sum += amount;
+
+      if (sum > totalFunds) {
+        revert InvalidWithdrawal();
+      }
+
+      totalAmount -= amount;
+
+      unchecked {
+        index++;
+      }
+    }
+
+    for (uint256 index = 0; index < amountLength; ) {
+      uint256 amount = amounts[index];
```

Comment:
can't this go inside the other loop? either way if we try to transfer more than available it's going to fail